horovod/common/elastic.py

```diff
@@ -149,17 +157,23 @@ def run_fn(func, reset):
     def wrapper(state, *args, **kwargs):
         notification_manager.init()
         notification_manager.register_listener(state)
+        skip_sync = False
 
         try:
             while True:
-                state.sync()
+                if not skip_sync:
+                    state.sync()
 
                 try:
                     return func(state, *args, **kwargs)
                 except HorovodInternalError:
                     state.restore()
-                except HostsUpdatedInterrupt:
-                    pass
+                    skip_sync = False
+                except HostsUpdatedInterrupt as e:
```

Comment:
We should be able to avoid `isinstance` here by putting the `HostsShrinkUpdatedInterrupt` catch above `HostsUpdatedInterrupt`.

Alternatively, we can only have `HostsUpdatedInterrupt`, but add an additional param: `skip_sync`.  Then we just have:

```
except HostsUpdatedInterrupt as e:
    skip_sync = e.skip_sync
```

I think this is cleaner.